Packs/CortexAttackSurfaceManagement/Scripts/InferWhetherServiceIsDev/InferWhetherServiceIsDev_test.py

```diff
@@ -9,47 +9,105 @@ def test_canonicalize():
+                          [{"Key": "ENV", "Value": "non-prd"}]),
+                          ([{"Key": "ENV", "Value": "prd"}], []),
+                             ([{"Key": "ENV", "Value": "dv"}, {"Key": "stage", "Value": "sbx"}],
+                              [{"Key": "ENV", "Value": "dv"}, {"Key": "stage", "Value": "sbx"}])
+                          ])
+def test_get_indicators_from_key_value_pairs(tags_raw, matches):
+    from InferWhetherServiceIsDev import get_indicators_from_key_value_pairs
+    from InferWhetherServiceIsDev import is_dev_indicator
+
+    assert get_indicators_from_key_value_pairs(tags_raw, is_dev_indicator) == matches
+
+
+def test_is_dev_indicator():
+    from InferWhetherServiceIsDev import is_dev_indicator
+
+    # Test Dev Matches
+    assert is_dev_indicator('dev')
+    assert is_dev_indicator('uat')
+    assert is_dev_indicator('non-prod')
+    assert is_dev_indicator('noprod')
+
+    # Test no match
+    assert not is_dev_indicator('devops')
+    assert not is_dev_indicator('prod')
+    assert not is_dev_indicator('pr')
+
+
+def test_is_prod_indicator():
+    from InferWhetherServiceIsDev import is_prod_indicator
+
+    # Test Dev Matches
+    assert is_prod_indicator('pr')
+    assert is_prod_indicator('prod')
+
+    # Test no Matches
+    assert not is_prod_indicator('non-prod')
+    assert not is_prod_indicator('staging')
+
+
+@pytest.mark.parametrize('classifications, matches', [(["SshServer", "DevelopmentEnvironment"],
+                                                       ["DevelopmentEnvironment"]),
+                                                      (["SshServer"], [])])
+def test_get_indicators_from_external_classification(classifications, matches):
+    from InferWhetherServiceIsDev import get_indicators_from_external_classification
+
+    assert get_indicators_from_external_classification(classifications) == matches
+
+
+@pytest.mark.parametrize('external, internal, reason',
+                         [(["DevelopmentEnvironment"], [],
+                          "Match on external classification of DevelopmentEnvironment"),
+                          (["DevelopmentEnvironment"], [{"Key": "env", "Value": "non-prod", "Source": "AWS"}],
+                          "Match on external classification of DevelopmentEnvironment and tag {env: non-prod} from AWS"),
+                          ([], [{"Key": "env", "Value": "non-prod", "Source": "AWS"}],
+                          "Match on tag {env: non-prod} from AWS"),
+                          ([], [{"Key": "env", "Value": "non-prod", "Source": "AWS"},
+                                {"Key": "stage", "Value": "sbx", "Source": "GCP"}],
+                          "Match on tag {env: non-prod} from AWS and tag {stage: sbx} from GCP")])
+def test_determine_reason(external, internal, reason):
+    from InferWhetherServiceIsDev import determine_reason
+
+    assert determine_reason(external, internal) == reason
+
+
```

Comment:
trying this in https://github.com/demisto/content/pull/28269/commits/322f2b01cb57f692360b9d9b0b07a6c276323f8b

Reply:
@ptoman-pa , does below look good for the truth table?

Reply:
no -- it only has 7 elements.

I had to reformat it to tell (I'd recommend not using the fixture and/or making each possible element into a constant -- fixtures are too freaking hard to read when the formats are complex):

```
# kv pair contains no indicators, DevEnv is set --> Likely Development
(["DevelopmentEnvironment"], [], [], {"reason": "Match on external classification of DevelopmentEnvironment", "result": True, "confidence": "Likely Development"}),

# kv pair contains no indicators, DevEnv is not set --> Not enough info
([], [], [], {"confidence": "Not Enough Information", "reason": "Neither dev nor prod indicators found", "result": False}),

# kv pair contains dev only, DevEnv is not set --> Likely Development
([], [{"Key": "env", "Value": "non-prod", "Source": "AWS"}], [], {"confidence": "Likely Development", "reason": "Match on tag {env: non-prod} from AWS", "result": True}),

# kv pair contains dev only, DevEnv is set --> Likely Development
(["DevelopmentEnvironment"], [{"Key": "env", "Value": "non-prod", "Source": "AWS"}], [], {"confidence": "Likely Development", "result": True, "reason": "Match on external classification of DevelopmentEnvironment and tag {env: non-prod} from AWS"}),

# kv pair contains prod only, DevEnv is set --> Conflicting
(["DevelopmentEnvironment"], [], [{"Key": "env", "Value": "prd", "Source": "AWS"}], {"result": False, "confidence": "Conflicting Information", "reason": "Match on external classification of DevelopmentEnvironment and tag {env: prd} from AWS"}),

# kv pair contains prod only, DevEnv is not set --> Likely Production
([], [], [{"Key": "env", "Value": "prd", "Source": "AWS"}],{"result": False, "confidence": "Likely Production", "reason": "Match on tag {env: prd} from AWS"}),

# kv pair contains dev and prod, DevEnv is not set --> Conflicting
([], [{"Key": "env", "Value": "dv", "Source": "Tenable.io"}], [{"Key": "env", "Value": "pr", "Source": "AWS"}], {"result": False, "confidence": "Conflicting Information", "reason": "Match on tag {env: dv} from Tenable.io and tag {env: pr} from AWS"})]
```

What is missing is:
```
kv pair contains dev and prod, DevEnv is set --> Conflicting
```

Reply:
@ptoman-pa , please see [ix tests](https://github.com/demisto/content/pull/28269/commits/eb5b48bbebfe15cbed499ed9f7142abc64b2e857)

Tests like 109

Reply:
Additional comment, @ptoman-pa I think this is harder to read on the PR but I feel your sentiment, I would use a dictionary here if that helps, I would advise against extrapolating out this many test for just different parameters.

```
reason_data = {
  "external": {
    "none": [],
    "non_prod": ["DevelopmentEnvironment"],
    "prod":["ProductionEnvironment"]
    },
  "internal": {
    "non_prod": [{"Key": "env", "Value": "non-prod", "Source": "AWS"}],
    "prod": [{"Key": "env", "Value": "prod", "Source": "AWS"}],
    "unknown_env_case": [{"Key": "env", "Value": "custom-tag", "Source": "AWS"}],
    "multiple_prod":[{"Key": "env", "Value": "non-prod", "Source": "AWS"}, {"Key": "stage", "Value": "sbx", "Source": "GCP"}]
  }
}

@pytest.mark.parametrize('external, internal', 'reason'
                         [(reason_data["external"]["non_prod"], reason_data["internal"]["non_prod"], "match on external classification of DevelopmentEnvironment and tag {env: non-prod} from AWS"))])
```

Reply:
Also @ptoman-pa shouldn't the test you mention for all elements be in `test_final_decision` not `test_determine_reason`. You could ignore `test_determine_reason` all together since `test_final_decision` covers it.

It looks like the `test_final_decision` should be using the full truthy table that you describe.